File: tools/memory/searchLongTermMemory.js
/**
 * Tool: Search Long-Term Memory
 * Searches through long-term memory
 */
export default {
  name: 'searchLongTermMemory',
  description: 'Search through long-term memory',
  parameters: { query: 'string' },
  
  async execute(params, context) {
    const { query } = params;
    const { longTermMemory } = context;
    
    const queryLower = query.toLowerCase();
    const results = longTermMemory.entries.filter(entry => {
      const keyMatch = entry.key.toLowerCase().includes(queryLower);
      const valueMatch = JSON.stringify(entry.value).toLowerCase().includes(queryLower);
      return keyMatch || valueMatch;
    });
    
    return {
      success: true,
      query,
      results: results.map(e => ({
        key: e.key,
        value: e.value,
        metadata: e.metadata,
        historyCount: e.history ? e.history.length : 0
      })),
      count: results.length
    };
  }
};

File: obscurex/tools/memory/search_longterm_memory.py
"""
Tool: Search Long-Term Memory
Searches through long-term memory
"""

from typing import Dict, Any


async def execute(params: Dict[str, Any], context: Dict[str, Any]) -> Dict[str, Any]:
    """Search through long-term memory."""
    query = params.get('query', '').lower()
    long_term_memory = context['longTermMemory']
    
    results = []
    for entry in long_term_memory['entries']:
        # Search in key, value (if string), and metadata
        key_match = query in str(entry['key']).lower()
        value_match = query in str(entry['value']).lower()
        metadata_match = query in str(entry.get('metadata', {})).lower()
        
        if key_match or value_match or metadata_match:
            # Don't include full history in search results
            result_entry = {
                'key': entry['key'],
                'value': entry['value'],
                'metadata': entry['metadata'],
                'historyCount': len(entry.get('history', []))
            }
            results.append(result_entry)
    
    return {
        'success': True,
        'query': query,
        'count': len(results),
        'results': results
    }


# Tool definition
tool = {
    'name': 'searchLongTermMemory',
    'description': 'Search through long-term memory',
    'parameters': {'query': 'string'},
    'execute': execute
}
